### ISSUE TITLE
Tracking grad norm without clutter

### ISSUE THREAD

**Issue by thegyro** [PyTorchLightning/pytorch-lightning#1462]
#### What is your question?
I am tracking my model's gradient norms by setting the flag `track_grad_norm=2`. However, this logs the individual norms of all the gradients. So my tensorboard is just filled with grad_norms before I can get to my training metrics. Mostly, I am interested only in the _total_ gradient norm. Is it possible to log just that? Logging all the gradients is fine if they can neatly be rolled into one directory in tensorboard.

**Comment by moi90** [comment]
Did you find a solution to your problem?

**Comment by cuican1432** [comment]
I have the same question

**Comment by perschi** [comment]
If pytorch-lightning is imported like this the function creating the norm dictionary that is logged can be decorated to create some structure. For that, I prepared two functions that decorate the function to adapt the dictionary.


In this first version every norm is removed from the dictionary except the total norm.

```
import pytorch_lightning as pl

def tracked_gradient_global_only():
    def remove_per_weight_norms(func):
        def f(*args):
            norms = func(*args)
            
            norms= dict(filter(lambda elem: '_total' in elem[0], norms.items()))
            
            return norms
        return f
            
    pl.core.grads.GradInformation.grad_norm = remove_per_weight_norms(pl.core.grads.GradInformation.grad_norm)
    
````

This second version adds the Tensorboard  subcategory 'gradients/'  to the dictionary key. Hence, all gradients can now be collapsed if not needed.

````
import pytorch_lightning as pl

def tracked_gradient_subcategory():
    def add_category(func):
        def f(*args):
            norms = func(*args)
            
            norms = {'gradients/' + k : v for k,v in norms.items()}
            print(norms)
            return norms
        return f
            
    pl.core.grads.GradInformation.grad_norm = add_category(pl.core.grads.GradInformation.grad_norm)
````

To use these snippets just copy the imports and the wanted adaption and call the function.

```
tracked_gradient_global_only()
#tracked_gradient_subcategory()
```

Just as a warning, I do not know if there are any side effects with any other functionalities, but for just training a simple model it works fine.

**Comment by Shikhar-S** [comment]
Has anyone found a solution to this? 

As a workaround I have been using the regex `^(?!grad_2.0)` to filter out the scalars I log with `self.log( )` .

**Comment by bmahlbrand** [comment]
> If pytorch-lightning is imported like this the function creating the norm dictionary that is logged can be decorated to create some structure. For that, I prepared two functions that decorate the function to adapt the dictionary.
> 
> In this first version every norm is removed from the dictionary except the total norm.
> 
> ```
> import pytorch_lightning as pl
> 
> def tracked_gradient_global_only():
>     def remove_per_weight_norms(func):
>         def f(*args):
>             norms = func(*args)
>             
>             norms= dict(filter(lambda elem: '_total' in elem[0], norms.items()))
>             
>             return norms
>         return f
>             
>     pl.core.grads.GradInformation.grad_norm = remove_per_weight_norms(pl.core.grads.GradInformation.grad_norm)
>     
> ```
> 
> This second version adds the Tensorboard subcategory 'gradients/' to the dictionary key. Hence, all gradients can now be collapsed if not needed.
> 
> ```
> import pytorch_lightning as pl
> 
> def tracked_gradient_subcategory():
>     def add_category(func):
>         def f(*args):
>             norms = func(*args)
>             
>             norms = {'gradients/' + k : v for k,v in norms.items()}
>             print(norms)
>             return norms
>         return f
>             
>     pl.core.grads.GradInformation.grad_norm = add_category(pl.core.grads.GradInformation.grad_norm)
> ```
> 
> To use these snippets just copy the imports and the wanted adaption and call the function.
> 
> ```
> tracked_gradient_global_only()
> #tracked_gradient_subcategory()
> ```
> 
> Just as a warning, I do not know if there are any side effects with any other functionalities, but for just training a simple model it works fine.

Something like this would be great to have in lightning properly

**Comment by HaukurPall** [comment]
Putting it all together for version 2.4:
```
class LogGradientNorm(Callback):
    """
    Logs the gradient norm.
    This should log the gradient norm unscaled and before clipping.
    """

    def on_before_optimizer_step(self, trainer: Trainer, *args, **kwargs) -> None:
        total_norm = 0.0
        for param in trainer.lightning_module.parameters():
            if param.grad is not None:
                param_norm = param.grad.detach().data.norm(2)
                total_norm += param_norm.item() ** 2
            total_norm = total_norm ** (1.0 / 2)
        trainer.lightning_module.log_dict({"train/grad_norm": total_norm})
```

**Comment by ronuchit** [comment]
For anyone trying to copy @HaukurPall 's code in the previous comment, note that there is an error: the line `total_norm = total_norm ** (1.0 / 2)` should be un-indented by one, so that it's outside the for loop.

**Comment by HaukurPall** [comment]
> For anyone trying to copy @HaukurPall 's code in the previous comment, note that there is an error: the line `total_norm = total_norm ** (1.0 / 2)` should be un-indented by one, so that it's outside the for loop.

Hi, thanks for pointing that out - I've fixed the bug in the snippet above

**Comment by kikatuso** [comment]
even easier implementation than's @HaukurPall 


```python
from lightning.pytorch.utilities import grad_norm

    def on_before_optimizer_step(self, optimizer):
        norm_order = 2.0 
        norms = grad_norm(self, norm_type=norm_order)
        self.log('Total gradient (norm)',norms[f'grad_{norm_order}_norm_total'],on_step=True, on_epoch=False)
```

**Comment by LucienXian** [comment]
> even easier implementation than's [@HaukurPall](https://github.com/HaukurPall)
> 
> from lightning.pytorch.utilities import grad_norm
> 
>     def on_before_optimizer_step(self, optimizer):
>         norm_order = 2.0 
>         norms = grad_norm(self, norm_type=norm_order)
>         self.log('Total gradient (norm)',norms[f'grad_{norm_order}_norm_total'],on_step=True, on_epoch=False)

If uses strategy = deepspeed_stage_3, grad_norm will return None. Is there any way to collect grad_norm under this premise?
